stellar-contract-env-common/src/env_val.rs

```diff
@@ -231,6 +252,98 @@ impl<E: Env> IntoEnvVal<E, RawVal> for u64 {
     }
 }
 
+#[cfg(feature = "std")]
+impl<E: Env + FromToXdrObj> TryFrom<EnvVal<E, RawVal>> for ScVal {
+    type Error = ConversionError;
+
+    fn try_from(ev: EnvVal<E, RawVal>) -> Result<Self, Self::Error> {
+        let env = ev.env;
+        let val = ev.val;
+        if val.is_u63() {
+            Ok(ScVal::U63(unsafe { val.unchecked_as_u63() }))
+        } else {
+            match val.get_tag() {
+                Tag::U32 => Ok(ScVal::U32(unsafe {
+                    <u32 as RawValConvertible>::unchecked_from_val(val)
+                })),
+                Tag::I32 => Ok(ScVal::I32(unsafe {
+                    <i32 as RawValConvertible>::unchecked_from_val(val)
+                })),
+                Tag::Static => {
+                    let tag_static =
+                        unsafe { <Static as RawValConvertible>::unchecked_from_val(val) };
+                    if tag_static.is_type(ScStatic::True) {
+                        Ok(ScVal::Static(ScStatic::True))
+                    } else if tag_static.is_type(ScStatic::False) {
+                        Ok(ScVal::Static(ScStatic::False))
+                    } else if tag_static.is_type(ScStatic::Void) {
+                        Ok(ScVal::Static(ScStatic::Void))
+                    } else if tag_static.is_type(ScStatic::LedgerKeyContractCodeWasm) {
+                        Ok(ScVal::Static(ScStatic::LedgerKeyContractCodeWasm))
+                    } else {
+                        Err(ConversionError)
```

Comment:
Unfortunately because these conversion can be used outside the host they can no longer return nice meaningful host errors. I don't see a way around this, so I'm going to press ahead with this, and maybe we can improve it somehow. Lmk if you see a way.

cc @graydon

Reply:
I put a lot of machinery for richer errors in EnvBase over in #243, maybe it's usable here?